src/Ops.jl

```diff
@@ -0,0 +1,430 @@
+                    result=mlir_type(TracedRArray{T,N}, size(a)),
+                    location,
+                ),
+            )
+            return TracedRArray{T,N}((), res, size(a))
+        end
+
+        function $op(
+            a::TracedRNumber{T},
+            b::TracedRNumber{T};
+            location=MLIR.IR.Location(
+                string($op), MLIR.IR.Location(@__FILE__, @__LINE__, 0)
+            ),
+        ) where {T}
+            res = MLIR.IR.result(
+                $op(
+                    a.mlir_data,
+                    b.mlir_data;
+                    result=mlir_type(TracedRArray{T,0}, ()),
+                    location,
+                ),
+            )
+            return TracedRNumber{T}((), res)
+        end
+    end
+end
+
+# is* checks
+for op in [:(stablehlo.is_finite), :(chlo.is_inf), :(chlo.is_neg_inf), :(chlo.is_pos_inf)]
+    @eval begin
+        function $op(
+            x::TracedRArray{T,N};
+            location=MLIR.IR.Location(
+                string($op), MLIR.IR.Location(@__FILE__, @__LINE__, 0)
+            ),
+        ) where {T,N}
+            res = MLIR.IR.result(
+                $op(x.mlir_data; result=mlir_type(TracedRArray{Bool,N}, size(x)), location)
+            )
+            return TracedRArray{Bool,N}((), res, size(x))
+        end
+
+        function $op(
+            x::TracedRNumber{T};
+            location=MLIR.IR.Location(
+                string($op), MLIR.IR.Location(@__FILE__, @__LINE__, 0)
+            ),
+        ) where {T}
+            res = MLIR.IR.result(
+                $op(x.mlir_data; result=mlir_type(TracedRArray{Bool,0}, ()), location)
+            )
+            return TracedRNumber{Bool}((), res)
+        end
+    end
+end
+
+# fixes to default automated implementations
+function stablehlo.abs(
+    x::TracedRArray{Complex{T},N};
+    location=MLIR.IR.Location("stablehlo.abs", MLIR.IR.Location(@__FILE__, @__LINE__, 0)),
+) where {T,N}
+    res = MLIR.IR.result(
+        stablehlo.abs(x.mlir_data; result=mlir_type(TracedRArray{T,N}, size(x)), location)
+    )
+    return TracedRArray{T,N}((), res, size(x))
+end
+
+function stablehlo.abs(
+    x::TracedRNumber{Complex{T}};
+    location=MLIR.IR.Location("stablehlo.abs", MLIR.IR.Location(@__FILE__, @__LINE__, 0)),
+) where {T}
+    res = MLIR.IR.result(
+        stablehlo.abs(x.mlir_data; result=mlir_type(TracedRArray{T,0}, ()), location)
+    )
+    return TracedRNumber{T}((), res)
+end
+
+# miscelaneous
+function stablehlo.complex(
+    real::TracedRArray{T,N},
+    imag::TracedRArray{T,N};
+    location=MLIR.IR.Location(
+        "stablehlo.complex", MLIR.IR.Location(@__FILE__, @__LINE__, 0)
+    ),
+) where {T,N}
+    res = MLIR.IR.result(
+        stablehlo.complex(
+            real.mlir_data,
+            imag.mlir_data;
+            result=mlir_type(TracedRArray{T,N}, size(real)),
+            location,
+        ),
+    )
+    return TracedRArray{T,N}((), res, size(real))
```

Comment:
I think this is missing

```suggestion
    return TracedRArray{Complex{T},N}((), res, size(real))
```